rust/datafusion/src/optimizer/projection_push_down.rs

```diff
@@ -0,0 +1,377 @@
+
+                // push projection down
+                let input = self.optimize_plan(&input, accum, mapping)?;
+
+                // rewrite filter expression to use new column indexes
+                let new_expr = self.rewrite_expr(expr, mapping)?;
+
+                Ok(Rc::new(LogicalPlan::Selection {
+                    expr: new_expr,
+                    input,
+                }))
+            }
+            LogicalPlan::Aggregate {
+                input,
+                group_expr,
+                aggr_expr,
+                schema,
+            } => {
+                // collect all columns referenced by grouping and aggregate expressions
+                self.collect_exprs(&group_expr, accum);
+                self.collect_exprs(&aggr_expr, accum);
+
+                // push projection down
+                let input = self.optimize_plan(&input, accum, mapping)?;
+
+                // rewrite expressions to use new column indexes
+                let new_group_expr = self.rewrite_exprs(group_expr, mapping)?;
+                let new_aggr_expr = self.rewrite_exprs(aggr_expr, mapping)?;
+
+                Ok(Rc::new(LogicalPlan::Aggregate {
+                    input,
+                    group_expr: new_group_expr,
+                    aggr_expr: new_aggr_expr,
+                    schema: schema.clone(),
+                }))
+            }
+            LogicalPlan::Sort {
+                expr,
+                input,
+                schema,
+            } => {
+                // collect all columns referenced by sort expressions
+                self.collect_exprs(&expr, accum);
+
+                // push projection down
+                let input = self.optimize_plan(&input, accum, mapping)?;
+
+                // rewrite sort expressions to use new column indexes
+                let new_expr = self.rewrite_exprs(expr, mapping)?;
+
+                Ok(Rc::new(LogicalPlan::Sort {
+                    expr: new_expr,
+                    input,
+                    schema: schema.clone(),
+                }))
+            }
+            LogicalPlan::EmptyRelation { schema } => {
+                Ok(Rc::new(LogicalPlan::EmptyRelation {
+                    schema: schema.clone(),
+                }))
+            }
+            LogicalPlan::TableScan {
+                schema_name,
+                table_name,
+                schema,
+                ..
+            } => {
+                // once we reach the table scan, we can use the accumulated set of column indexes as
+                // the projection in the table scan
+                let mut projection: Vec<usize> = Vec::with_capacity(accum.len());
+                accum.iter().for_each(|i| projection.push(*i));
+
+                // sort the projection otherwise we get non-deterministic behavior
+                projection.sort();
+
+                // now that the table scan is returning a different schema we need to create a
```

Comment:
Why is it "returning a different schema" - seems the same schema is still returned?

Also, to help me understand, does each plan operator has its own schema and it could be different from the global schema (e.g., the schema of the input source). If so, is the column index the index to the schema of the **current** schema (e.g., column index in the `expr` of `Sort` will point to the `schema` in the `Sort` plan)?

Reply:
Looks like you just found a bug. Good catch. It should be returning the schema after the projection has been applied.

Yes each plan operator has its own schema (for its output). In some cases (filter, limit, sort) the schema does not change so they can just delegate to their input relation.

Column indexes are always for the schema of the input relation.

Reply:
I pushed a fix for the bug and added a test

Reply:
> Column indexes are always for the schema of the input relation.

Does the schema of the input relation change? I'm still not sure why we need to rewrite the column indexes - can they always point to the complete schema of the input source?

Reply:
The logical plan created via the SQL query planner (and the DataFrame API when we have one) does just refer to the original table schema.

The query optimizer transforms the plan and pushes the projection down to the TableScan so that we basically pretend the table only contains the columns we care about. The rest of the plan is then rewritten to be relative to that.

Each operator in the plan is relative to its input and doesn't know about the underlying table schema which could be many levels down, especially once we have joins and subqueries.

Reply:
I guess one of the reasons for doing this, other than having a concise and simple to comprehend plan, is that ultimately these are indexes into RecordBatch instances. 

Let's say we have a csv/parquet file with 300 columns and the query only references 12 of them... If we don't do this rewriting then we are going to have to load a RecordBatch with 300 columns where 288 of them are empty arrays / or empty options, or we have to have some special implementation of RecordBatch which does a mapping.

Reply:
OK. Makes sense. Thanks.